### PR TITLE
ADX-526 Send email notifications for dataset transfer requests

### PR DESCRIPTION
**Pull request by Manoj-nathwani**
- Added `IPackageController after_update` to unaids
- Created a `DatasetTransferRequest` model to track who we've emailed for every `dataset_id <> org_id` instance
- See https://github.com/fjelltopp/adx_develop/pull/36
- `ckanext/unaids/command.py` and `setup.py` has been blindly copied from ytp-request with small tweaks

### CODE REVIEW

File: ckanext/unaids/dataset_transfer/logic.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+from ckanext.unaids.dataset_transfer.model import (
+    DatasetTransferRequest,
+    STATUS_EMAILED,
+    STATUS_EMAIL_FAILED
+)
+
+log = logging.getLogger(__name__)
+
+
+def send_dataset_transfer_emails(dataset_id, recipient_org_id):
```
Comment (tomeksabala) [inline ckanext/unaids/dataset_transfer/logic.py:15]: This method is too long and too complex. You should extract some smaller pieces of logic into separate methods.
That's going to improve the readability and more importantly you'll be able to create unit tests for those smaller part of logic easily.

**Review by tomeksabala — CHANGES_REQUESTED** [review]
1. Create the tests ASAP :) !!!!
2. Some files doesn't end with a new line. Please fix this.
3. I'd need to take a closer look at sqlalchemy configuration for unaids extension. I've never done it before. I think we could talk about it later separately.
4. I like how you've created the after_update method. It's good that we're fetching the `org_to_allow_transfer_to` there as it's part of the dataset configuration. At the same time the API of your `dataset_transfer` looks good.

**Comment by Manoj-nathwani** [comment]
# Update:
- Added tests
- This is 99% complete now.
- I just need to add tests checking the html body
- and generally make sure I have good enough test coverage
- otherwise I think i'm done here

@tomeksabala feedback would be appreciated :)

File: ckanext/unaids/plugin.py [hunk 2]
```diff
@@ -119,6 +120,19 @@ def can_validate(self, context, data_dict):
         if data_dict.get('schema'):
             return True
 
+    # IPackageController
+    def after_update(self, context, pkg_dict):
+        org_to_allow_transfer_to = [
+            item['value']
+            for item in pkg_dict['extras']
+            if item['key'] == 'org_to_allow_transfer_to'
+        ][0]
```
Comment (Manoj-nathwani) [inline ckanext/unaids/plugin.py:129]: ❗ This is causing a bug when uploading resources with

```
Traceback (most recent call last):
  File "/usr/lib/ckan/venv/lib/python2.7/site-packages/flask/app.py", line 2449, in wsgi_app
    response = self.handle_exception(e)
  File "/usr/lib/ckan/venv/lib/python2.7/site-packages/flask/app.py", line 1866, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/lib/ckan/venv/lib/python2.7/site-packages/flask/app.py", line 2446, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/lib/ckan/venv/lib/python2.7/site-packages/flask/app.py", line 1951, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/lib/ckan/venv/lib/python2.7/site-packages/flask/app.py", line 1820, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/lib/ckan/venv/lib/python2.7/site-packages/flask/app.py", line 1949, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/lib/ckan/venv/lib/python2.7/site-packages/flask_debugtoolbar/__init__.py", line 125, in dispatch_request
    return view_func(**req.view_args)
  File "/usr/lib/ckan/venv/lib/python2.7/site-packages/flask/views.py", line 89, in view
    return self.dispatch_request(*args, **kwargs)
  File "/usr/lib/ckan/venv/lib/python2.7/site-packages/flask/views.py", line 163, in dispatch_request
    return meth(*args, **kwargs)
  File "/usr/lib/ckan/venv/src/ckan/ckan/views/resource.py", line 248, in post
    get_action(u'resource_create')(context, data)
  File "/usr/lib/ckan/venv/src/ckan/ckan/logic/__init__.py", line 473, in wrapped
    result = _action(context, data_dict, **kw)
  File "/usr/lib/ckan/venv/src/ckan/ckan/logic/action/create.py", line 319, in resource_create
    _get_action('package_update')(context, pkg_dict)
  File "/usr/lib/ckan/venv/src/ckan/ckan/logic/__init__.py", line 473, in wrapped
    result = _action(context, data_dict, **kw)
  File "/usr/lib/ckan/venv/src/ckan/ckan/logic/action/update.py", line 330, in package_update
    item.after_update(context, data)
  File "/usr/lib/adx/ckanext-unaids/ckanext/unaids/plugin.py", line 129, in after_update
    ][0]
IndexError: list index out of range
```

Comment (Manoj-nathwani) [reply ckanext/unaids/plugin.py]: Fixed in https://github.com/fjelltopp/ckanext-unaids/pull/92/commits/ad8981aa417930637e53f7394049f335e9ec5e8f

**Comment by Manoj-nathwani** [comment]
All done ✅ 

Here's an example of what the email looks like:
```
Dear test_user_01,
         
The test_org_00 administrator has requested a transfer of dataset  to your organisation test_org_01.
You can view the dataset with the following link: http://test.ckan.net/packages?id=test_dataset_00

If you would like to accept the transfer click the "Transfer dataset to test_org_01" while viewing the dataset.
Or accept the transfer immediately by clicking the link:
http://test.ckan.net/validation/transfer_dataset/test_dataset_00

Best regards,
AIDS Data Repository Administrator
```

**Comment by Manoj-nathwani** [comment]
Fixed bugs with urls ✅ 
```
Dear test_user_01,

The test_org_00 administrator has requested a transfer of dataset  to your organisation test_org_01.
You can view the dataset with the following link: http://test.ckan.net/dataset/test_dataset_00

If you would like to accept the transfer click the "Transfer dataset to test_org_01" while viewing the dataset.
Or accept the transfer immediately by clicking the link:
http://test.ckan.net/validation/transfer_dataset/2f8644f4-4408-4947-bf51-783b462eda7b

Best regards,
AIDS Data Repository Administrator
```

File: ckanext/unaids/theme/templates/email/dataset_transfer.html [hunk 3]
```diff
@@ -0,0 +1,11 @@
+Dear {{ user.name }},
```
Comment (tomeksabala) [inline ckanext/unaids/theme/templates/email/dataset_transfer.html:1]: This should be a *.txt file not *.html

File: ckanext/unaids/dataset_transfer/logic.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                    'dataset_org': dataset_org,
+                    'recipient_org': recipient_org,
+                    'dataset_url': config.get('ckan.site_url') + dataset_url,
+                    'accept_transfer_url': config.get('ckan.site_url') + accept_transfer_url
+                }
+            )
+            mailer.mail_user(user, subject, body)
+            status = STATUS_EMAILED
+        except mailer.MailerException as e:
+            log.debug(e.message)
```
Comment (tomeksabala) [inline ckanext/unaids/dataset_transfer/logic.py:85]: We should log this exception as an error here (maybe without a stack trace).

File: ckanext/unaids/tests/test_dataset_transfer.py [hunk 5]
```diff
@@ -1,16 +1,83 @@
...
+            recipient_org_id=org_2['id']
+        )
+        mocked_mail_user.assert_called()
+        emailed_user_ids = [
+            x[0][0].id  # recipient_user_id
+            for x in mocked_mail_user.call_args_list
+        ]
+        assert emailed_user_ids == [user_2['id']]
+
+    def test_get_users_to_email(self, app):
```
Comment (tomeksabala) [inline ckanext/unaids/tests/test_dataset_transfer.py:55]: This is an all-in-one test case. Please create a separate test case for each of a use case of the function under test.

**Review by tomeksabala — CHANGES_REQUESTED** [review]
A few changes in code recommended in inline comments. Also when I tested it locally I go the following error:
```
ckan          | Traceback (most recent call last):
ckan          |   File "/usr/lib/ckan/venv/lib/python2.7/site-packages/flask/app.py", line 2449, in wsgi_app
ckan          |     response = self.handle_exception(e)
ckan          |   File "/usr/lib/ckan/venv/lib/python2.7/site-packages/flask/app.py", line 1866, in handle_exception
ckan          |     reraise(exc_type, exc_value, tb)
ckan          |   File "/usr/lib/ckan/venv/lib/python2.7/site-packages/flask/app.py", line 2446, in wsgi_app
ckan          |     response = self.full_dispatch_request()
ckan          |   File "/usr/lib/ckan/venv/lib/python2.7/site-packages/flask/app.py", line 1951, in full_dispatch_request
ckan          |     rv = self.handle_user_exception(e)
ckan          |   File "/usr/lib/ckan/venv/lib/python2.7/site-packages/flask/app.py", line 1820, in handle_user_exception
ckan          |     reraise(exc_type, exc_value, tb)
ckan          |   File "/usr/lib/ckan/venv/lib/python2.7/site-packages/flask/app.py", line 1949, in full_dispatch_request
ckan          |     rv = self.dispatch_request()
ckan          |   File "/usr/lib/ckan/venv/lib/python2.7/site-packages/flask_debugtoolbar/__init__.py", line 125, in dispatch_request
ckan          |     return view_func(**req.view_args)
ckan          |   File "/usr/lib/ckan/venv/lib/python2.7/site-packages/flask/views.py", line 89, in view
ckan          |     return self.dispatch_request(*args, **kwargs)
ckan          |   File "/usr/lib/ckan/venv/lib/python2.7/site-packages/flask/views.py", line 163, in dispatch_request
ckan          |     return meth(*args, **kwargs)
ckan          |   File "/usr/lib/ckan/venv/src/ckan/ckan/views/dataset.py", line 734, in post
ckan          |     pkg_dict = get_action(u'package_update')(context, data_dict)
ckan          |   File "/usr/lib/ckan/venv/src/ckan/ckan/logic/__init__.py", line 473, in wrapped
ckan          |     result = _action(context, data_dict, **kw)
ckan          |   File "/usr/lib/ckan/venv/src/ckan/ckan/logic/action/update.py", line 330, in package_update
ckan          |     item.after_update(context, data)
ckan          |   File "/usr/lib/adx/ckanext-unaids/ckanext/unaids/plugin.py", line 133, in after_update
ckan          |     recipient_org_id=org_to_allow_transfer_to[0]
ckan          |   File "/usr/lib/adx/ckanext-unaids/ckanext/unaids/dataset_transfer/logic.py", line 59, in send_dataset_transfer_emails
ckan          |     exclude_user_ids=user_ids_already_emailed
ckan          |   File "/usr/lib/adx/ckanext-unaids/ckanext/unaids/dataset_transfer/logic.py", line 21, in _get_users_to_email
ckan          |     for user in recipient_org['users']
ckan          | KeyError: 'users'
```
Steps:
1. adx up && adx dbsetup && adx demodata
2. log in as admin/fjelltopp
3. Manage user-log-data dataset
4. Select transfer to North Pole
5. Click submit

**Comment by Manoj-nathwani** [comment]
Note for monday:
- tests for `send_dataset_transfer_emails` are failing with `Forbidden: 403 Forbidden: You cannot carry out this action`

File: ckanext/unaids/dataset_transfer/logic.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    STATUS_EMAIL_FAILED
+)
+
+log = logging.getLogger(__name__)
+
+
+def _get_users_to_email(recipient_org, exclude_user_ids):
+    recipient_org_admin_ids = [
+        user['id']
+        for user in recipient_org['users']
```
[the remark this answers is not in the document]
Comment (Manoj-nathwani) [reply ckanext/unaids/dataset_transfer/logic.py]: Were you thinking something like this?
Just want to double check as it seems a little ugly

```
assert 'users' in org, ValueError(' '.join([
    _('Organization'),
    org['name'],
    _('has no users to accept this transfer')
]))

org_admin_ids = [
    user['id']
    for user in org['users']
    if user['capacity'] == 'admin'
]
```

File: ckanext/unaids/dataset_transfer/logic.py (hunk 4 above)
Comment (Manoj-nathwani) [reply ckanext/unaids/dataset_transfer/logic.py:85]: Just to confirm, something like this?

```
log.error('Sending email for DatasetTransferRequest failed for user <USER_ID> dataset_id <DATASET_ID>')
log.error(e.message)
```

Comment (Manoj-nathwani) [reply ckanext/unaids/dataset_transfer/logic.py:85]: I guess we'll also want to add:

`h.flash_error(_('Failed to email organization admins. Please contact the dataset administrator manually'))`

File: ckanext/unaids/dataset_transfer/logic.py [hunk 7]
```diff
@@ -60,16 +60,16 @@ def send_dataset_transfer_emails(dataset_id, recipient_org_id):
...
     for user in users_to_email:
         try:
-            subject = 'testing123'
+            subject = _('Dataset Transfer')
             body = render_jinja2(
-                'email/dataset_transfer.html',
+                'email/dataset_transfer.txt',
                 extra_vars={
                     'user': user,
                     'dataset': 'dataset',
```
Comment (tomeksabala) [inline ckanext/unaids/dataset_transfer/logic.py:75]:
```
 'dataset': 'dataset',
```
How is this supposed to work? Jinja template is referencing `dataset.name` so the whole e-mail request is probably going to throw an template exception.

File: ckanext/unaids/dataset_transfer/logic.py [hunk 8]
```diff
@@ -60,16 +60,16 @@ def send_dataset_transfer_emails(dataset_id, recipient_org_id):
...
-        recipient_org=recipient_org,
+    users_to_email = get_org_admins_with_email_addresses(
+        org=recipient_org,
         exclude_user_ids=user_ids_already_emailed
     )
 
     for user in users_to_email:
         try:
-            subject = 'testing123'
+            subject = _('Dataset Transfer')
```
Comment (tomeksabala) [inline ckanext/unaids/dataset_transfer/logic.py:70]: Could you include the title of the dataset in the e-mail subject. That should be more informative.

File: ckanext/unaids/theme/templates/email/dataset_transfer.txt [hunk 9]
```diff
@@ -0,0 +1,12 @@
+Dear {{ user.name }},
+
+    The {{ dataset_org.name }} administrator has requested a transfer of dataset {{ dataset.name }} to your organisation {{ recipient_org.name }}.
```
Comment (tomeksabala) [inline ckanext/unaids/theme/templates/email/dataset_transfer.txt:3]: `dataset.name` is `a-slugified-title-of-a-dataset`. You sure you want to use it here? I think you've meant `dataset.title`.

**Review by tomeksabala — CHANGES_REQUESTED** [review]

File: ckanext/unaids/dataset_transfer/logic.py (hunk 8 above)
Comment (Manoj-nathwani) [reply ckanext/unaids/dataset_transfer/logic.py:70]: Sure, i'll make it `<DATASET_TITLE> Dataset Transfer`

File: ckanext/unaids/dataset_transfer/logic.py (hunk 7 above)
Comment (Manoj-nathwani) [reply ckanext/unaids/dataset_transfer/logic.py:82]: No no. If you send the `dataset` object to jinja2, then in jinja2 you can access the object like `dataset.name` without any issues.

File: ckanext/unaids/dataset_transfer/logic.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+        org_admin_ids = [
+            user['id']
+            for user in org['users']
+            if user['capacity'] == 'admin'
+        ]
+    except KeyError:
+        raise ValueError('{} {} {}'.format(
+            _('Organization'),
+            org['name'],
+            _('has no users to accept this transfer')
```
Comment (tomeksabala) [inline ckanext/unaids/dataset_transfer/logic.py:28]: The message for this ValueError shouldn't be translated. It's just internal implementation and shouldnt' surface to the UI.
Also the message is a little misleading. The function is about getting admins from an organization. There's no transfer of ownership in this context. The message should indicate how to use it properly and what went wrong, e.g.
"Provided organization {org.get(id)} doesn't include information about users. organization('users') = {org.get('users')}"

**Review by tomeksabala — CHANGES_REQUESTED** [review]

File: ckanext/unaids/plugin.py [hunk 11]
```diff
@@ -127,6 +127,7 @@ def after_update(self, context, pkg_dict):
                 item['value']
                 for item in pkg_dict['extras']
                 if item['key'] == 'org_to_allow_transfer_to'
+                and item['value']
```
[the remark this answers is not in the document]
Comment (Manoj-nathwani) [reply ckanext/unaids/plugin.py:130]: I think this is pretty pythonic and acceptable. I'd prefer to keep it
It's readable and clean-looking

File: ckanext/unaids/dataset_transfer/logic.py (hunk 10 above)
Comment (Manoj-nathwani) [reply ckanext/unaids/dataset_transfer/logic.py:28]: Yeah I wasn't sure about this as there isn't much consistency as to if we translate error messages 😅 
I'll change it as advised

**Comment by tomeksabala** [comment]
After playing around with the feature locally there are more TODOs:
1. Sys admin receives e-mails about every dataset transfer
2. With transfers: Fjelltopp > North Pole, North Pole > Fjelltopp, Fjelltopp > North Pole - there are no e-mails send with the 3rd transfer

**Comment by Manoj-nathwani** [comment]
> 2. With transfers: Fjelltopp > North Pole, North Pole > Fjelltopp, Fjelltopp > North Pole - there are no e-mails send with the 3rd transfer

This was a known feature. So you can only ever be emailed once regarding a dataset.
We said it was very very rare for a dataset to transfer to/from/to orgs, so there's no need to keep sending emails.

If we really want to keep emailing, then we could delete/expire all email communications whenever a dataset has been successfully transferred. This seems a bit messy though.

**Review by tomeksabala — APPROVED** [review]

**Pull request merged**